File: 5-Problem-A-Day/337. House Robber III.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public int rob(TreeNode root) {
        int[] ans = up(root);
        return Math.max(ans[0], ans[1]);
    }

    private int[] up(TreeNode root) {
        if (root == null)
            return new int[] {0, 0};
        int[] left = up(root.left);
        int[] right = up(root.right);
        return new int[] {Math.max(root.val + left[1] + right[1], left[0] + right[0]),
                left[0] + right[0]};
    }
}

File: 2021/337. House Robber III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        auto res = dfs(root);
        return max(res.first, res.second);
    }
    
    pair<int, int> dfs(TreeNode* root) {
        if (root == NULL) return {0, 0};
        auto left = dfs(root->left);
        auto right = dfs(root->right);        
        return {max(root->val + left.second + right.second, left.first + right.first), left.first + right.first};
    }
};